src/dmd/root/array.d

```diff
@@ -591,3 +591,81 @@ unittest
     static assert (!is(typeof(Array!(OtherStruct).init.sort())));
     static assert (!is(typeof(Array!(OtherStruct).init.sort!pred)));
 }
+
+/**
+ * Iterates the given array and calls the given callable for each element.
+ *
+ * Use this instead of `foreach` when the array may expand during iteration.
+ *
+ * Params:
+ *  callable = the callable to call for each element
+ *  array = the array to iterate
+ */
+template each(alias callable, T)
```

Comment:
Wouldn't this be better suited as a member method of `Array`, e.g. as `opApply`?

Reply:
This is a generalization of `foreachDsymbol`:

https://github.com/dlang/dmd/blob/ae95799c0821779d21b28be59eb152656059371c/src/dmd/dsymbol.d#L65

I assume that function is being used instead of `opApply` because of this from the spec:

"The aggregate must be loop invariant, meaning that elements to the aggregate cannot be added or removed from it in the NoScopeNonEmptyStatement." [1].

It technically works with `opApply`, but I'm not sure if that's allowed according to the spec.

[1] https://dlang.org/spec/statement.html#foreach-statement.

Reply:
> This is a generalization of `foreachDsymbol`:

Fair enough.

> "The aggregate must be loop invariant, meaning that elements to the aggregate cannot be added or removed from it in the NoScopeNonEmptyStatement." [1].

Guess that simplifies generic code (allthough this statement sounds more like a best practice rather than a language rule).
Nevermind then.

Reply:
From the PR that introduced `foreachDsymbol`:

"foreach can't be used because it caches the loop parameters, and there have been several bugs where the array expanded during iteration." [1].

I tried with implementing `opApply` and it works even when the array is expanded. But I'm not sure if Walter would like that.

[1] https://github.com/dlang/dmd/pull/9091#issuecomment-447723141